origin-dapp-creator-server/src/app.js

```diff
@@ -0,0 +1,114 @@
+const bodyParser = require('body-parser')
+const cors = require('cors')
+const express = require('express')
+const Logger = require('logplease')
+const Web3 = require('web3')
+const web3 = new Web3()
+
+const app = express()
+const port = process.env.PORT || 4321
+const logger = Logger.create('origin-dapp-creator-server')
+
+import { getDnsRecord, parseDnsTxtRecord, setAllRecords, updateTxtRecord } from './lib/dns'
+import { addConfigToIpfs, ipfsClient, getConfigFromIpfs } from './lib/ipfs'
+
+app.use(bodyParser.json())
+app.use(bodyParser.urlencoded({ extended: true }))
+app.use(cors())
+
+/* Route for handling creation of configurations. Saves the configuration to
+ * IPFS and configures a subdomain if necessary. Subdomains are protected via
+ * web3 signatures so it isn't possible to overwrite another users subdomain.
+ */
+app.post('/config', async (req, res) => {
+  const { config, signature, address } = req.body
+  let existingRecord
+
+  if (config.subdomain) {
+    // Validating signing is only necessary if we are configuring for a subdomain
+
+    // Validate signature matches
+    const signer = web3.eth.accounts.recover(JSON.stringify(config), signature)
+    // Address from recover is checksummed so lower case it
+    if (signer.toLowerCase() !== address) {
+      return res.status(400).send('Signature was invalid')
+    }
+
+    // Check if there is an existing configuration published for this subdomain
+    try {
+      existingRecord = await getDnsRecord(config.subdomain, 'TXT')
+    } catch (error) {
+      logger.error(error)
+      return res.status(500).send('An error occurred retrieving DNS records')
+    }
+
+    if (existingRecord) {
+      const ipfsHash = parseDnsTxtRecord(existingRecord.data[0])
+      if (!ipfsHash) {
+        return res.status(500).send('An error occurred retrieving existing configuration')
+      }
+      const existingConfig = await getConfigFromIpfs(ipfsHash)
+      if (existingConfig.address !== address) {
+        return res.status(400).send({
+          subdomain: 'Subdomain in use by another wallet'
+        })
+      }
+    }
+
+    logger.debug('Validated signature of configuration')
+  }
+
+  // Add the new config to IPFS
+  let ipfsHash
+  try {
+    ipfsHash = await addConfigToIpfs(req.body)
```

Comment:
Not as part of this PR, but maybe add a TODO for later to pin this content on the protocol labs cluster ?